File: src/features/backend/backends/localService/protocol.ts
/**
 * The Nova local service wire protocol, as the frontend speaks it.
 *
 * This module is pure: it defines the envelopes, the error taxonomy and the
 * limits, and it encodes and validates messages. It has no transport, clock or
 * randomness of its own, so every rule here is unit-tested without an
 * application, a service or an IPC bridge around it.
 *
 * The service's copy of these constants lives in
 * `src-tauri/src/local_service/protocol.rs`. Nothing at runtime could notice
 * the two drifting apart — a mismatched version would simply refuse every
 * request — so `protocolParity.test.ts` reads that file and compares them.
 *
 * Deliberately absent from every type here, as from the service itself: process
 * ids, process handles, memory addresses, module lists, executable paths and
 * evaluated code. The only secret that crosses this boundary is the session
 * token, and it appears in exactly one place — the handshake reply — and is
 * never put in a log line, a diagnostic or a persisted value.
 */

/** Bumped only on an incompatible change; both sides must agree on it exactly. */
export const LOCAL_SERVICE_PROTOCOL_VERSION = "NOVA_LOCAL_SERVICE_V1";

/** Largest message accepted in either direction, in bytes. */
export const LOCAL_SERVICE_MAX_MESSAGE_BYTES = 64 * 1024;

/** Every operation the service answers. */
export const LOCAL_SERVICE_OPS = ["handshake", "health", "attach", "detach", "shutdown"] as const;

export type LocalServiceOp = (typeof LOCAL_SERVICE_OPS)[number];

/** The one capability the service implements, and so the only one to require. */
export const LOCAL_SERVICE_CAPABILITY_TARGET = "target";

/**
 * Refusals the *service* produces. Each is the answer to exactly one rule, and
 * the names match `ErrorCode::as_str` in the Rust module one for one.
 */
export const LOCAL_SERVICE_WIRE_ERROR_CODES = [
  "MESSAGE_TOO_LARGE",
  "MESSAGE_MALFORMED",
  "PROTOCOL_UNSUPPORTED",
  "AUTH_REQUIRED",
  "AUTH_INVALID",
  "SESSION_STALE",
  "CAPABILITY_UNSUPPORTED",
  "OPERATION_UNSUPPORTED",
  "ALREADY_ATTACHED",
  "SERVICE_UNAVAILABLE",
] as const;

/**
 * Refusals the *client* produces, for the failures that never reach the
 * service. They are kept separate so a reader always knows which side decided:
 * a `TRANSPORT_FAILED` is Nova's own report about the bridge, while an
 * `AUTH_INVALID` is the service's answer.
 */
export const LOCAL_SERVICE_CLIENT_ERROR_CODES = [
  /** There is no IPC bridge: Nova is running in a plain browser. */
  "TRANSPORT_UNAVAILABLE",
  /** The IPC call itself failed. */
  "TRANSPORT_FAILED",
  /** The service did not answer within the request timeout. */
  "REQUEST_TIMEOUT",
  /** The answer was not a well-formed response envelope. */
  "RESPONSE_MALFORMED",
  /** The answer named a different request, so it belongs to nothing current. */
  "RESPONSE_MISMATCHED",
] as const;

export type LocalServiceWireErrorCode = (typeof LOCAL_SERVICE_WIRE_ERROR_CODES)[number];
export type LocalServiceClientErrorCode = (typeof LOCAL_SERVICE_CLIENT_ERROR_CODES)[number];
export type LocalServiceErrorCode = LocalServiceWireErrorCode | LocalServiceClientErrorCode;

export interface LocalServiceError {
  readonly code: LocalServiceErrorCode;
  /** Short sentence suitable for the UI. */
  readonly message: string;
  /** Technical context. Never a token, never a path. */
  readonly details?: string;
}

/** One request, as it goes on the wire. */
export interface LocalServiceRequest {
  readonly protocol: string;
  readonly op: LocalServiceOp;
  readonly requestId: string;
  /** Null only for a handshake, which is the operation that issues a session. */
  readonly sessionId: string | null;
  readonly token: string | null;
  readonly payload: Record<string, unknown>;
}

export type LocalServiceResult<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: LocalServiceError };

/** What the service says it can do. Every flag is reported, including the false ones. */
export interface LocalServiceCapabilities {
  readonly target: boolean;
  readonly debugger: boolean;
  readonly profiler: boolean;
  readonly execute: boolean;
}

/** What a handshake returns, minus the token, which never leaves the session client. */
export interface LocalServiceSessionInfo {
  readonly sessionId: string;
  readonly service: { readonly name: string; readonly version: string; readonly protocol: string };
  readonly capabilities: LocalServiceCapabilities;
  /** Epoch milliseconds, as the service measured it. */
  readonly openedAt: number;
  /** How many sessions the service has issued since it started. */
  readonly sessionsIssued: number;
}

/** What one health request found. */
export interface LocalServiceHealth {
  readonly target: { readonly available: boolean; readonly ready: boolean; readonly version: string };
  readonly attached: boolean;
  readonly capabilities: LocalServiceCapabilities;
  readonly sessionId: string | null;
  readonly attachedAt: number | null;
  readonly sessionsIssued: number;
}

const WIRE_CODES = new Set<string>(LOCAL_SERVICE_WIRE_ERROR_CODES);

const encoder = new TextEncoder();

export function byteLength(text: string): number {
  return encoder.encode(text).length;
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const isBoolean = (value: unknown): value is boolean => typeof value === "boolean";

/** Builds a well-formed request. */
export function createRequest(input: {
  op: LocalServiceOp;
  requestId: string;
  sessionId?: string | null;
  token?: string | null;
  payload?: Record<string, unknown>;
}): LocalServiceRequest {
  return {
    protocol: LOCAL_SERVICE_PROTOCOL_VERSION,
    op: input.op,
    requestId: input.requestId,
    sessionId: input.sessionId ?? null,
    token: input.token ?? null,
    payload: input.payload ?? {},
  };
}

export function encodeRequest(request: LocalServiceRequest): string {
  return JSON.stringify(request);
}

/**
 * Reads one answer.
 *
 * The order mirrors the service's own: size, then JSON, then the envelope
 * shape, then the protocol version, then the request it claims to answer. An
 * answer for a different request is refused rather than applied — that is the
 * client half of the rule that a stale reply can never affect a live request.
 *
 * It never throws: every failure is a structured {@link LocalServiceError}.
 */
export function decodeResponse(raw: string, requestId: string): LocalServiceResult<unknown> {
  if (byteLength(raw) > LOCAL_SERVICE_MAX_MESSAGE_BYTES) {
    return {
      ok: false,
      error: {
        code: "RESPONSE_MALFORMED",
        message: "The local service sent more than the protocol allows.",
        details: `the limit is ${LOCAL_SERVICE_MAX_MESSAGE_BYTES} bytes`,
      },
    };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw) as unknown;
  } catch {
    return {
      ok: false,
      error: { code: "RESPONSE_MALFORMED", message: "The local service sent something that is not JSON." },
    };
  }

  if (!isRecord(parsed)) {
    return {
      ok: false,
      error: { code: "RESPONSE_MALFORMED", message: "The local service sent something that is not a response." },
    };
  }

  if (parsed.protocol !== LOCAL_SERVICE_PROTOCOL_VERSION) {
    return {
      ok: false,
      error: {
        code: "PROTOCOL_UNSUPPORTED",
        message: "The local service speaks a different protocol version.",
        details: `expected ${LOCAL_SERVICE_PROTOCOL_VERSION}, received ${String(parsed.protocol).slice(0, 64)}`,
      },
    };
  }

  if (parsed.requestId !== requestId) {
    return {
      ok: false,
      error: {
        code: "RESPONSE_MISMATCHED",
        message: "The local service answered a different request.",
        details: "the answer was discarded",
      },
    };
  }

  if (parsed.ok === true) {
    if (!isRecord(parsed.payload)) {
      return {
        ok: false,
        error: { code: "RESPONSE_MALFORMED", message: "The local service sent an answer with no payload." },
      };
    }
    return { ok: true, value: parsed.payload };
  }

  if (parsed.ok !== false) {
    return {
      ok: false,
      error: { code: "RESPONSE_MALFORMED", message: "The local service sent an answer that is neither a result nor a refusal." },
    };
  }

  const error = readWireError(parsed.error);
  if (error === null) {
    return {
      ok: false,
      error: { code: "RESPONSE_MALFORMED", message: "The local service refused the request without saying why." },
    };
  }
  return { ok: false, error };
}

/** A refusal the service sent, or null when it is not one. */
function readWireError(value: unknown): LocalServiceError | null {
  if (!isRecord(value)) return null;
  const { code, message, details } = value;
  if (typeof code !== "string" || !WIRE_CODES.has(code)) return null;
  if (typeof message !== "string" || message === "") return null;
  return {
    code: code as LocalServiceWireErrorCode,
    message,
    ...(typeof details === "string" && details !== "" ? { details } : {}),
  };
}

function readCapabilities(value: unknown): LocalServiceCapabilities | null {
  if (!isRecord(value)) return null;
  const { target, debugger: debug, profiler, execute } = value;
  if (!isBoolean(target) || !isBoolean(debug) || !isBoolean(profiler) || !isBoolean(execute)) return null;
  return Object.freeze({ target, debugger: debug, profiler, execute });
}

/** Validates a handshake payload. The token is returned separately and is never part of the info. */
export function readSessionOpened(
  payload: unknown,
): { info: LocalServiceSessionInfo; token: string } | null {
  if (!isRecord(payload)) return null;
  const { sessionId, token, service, capabilities, openedAt, sessionsIssued } = payload;
  if (typeof sessionId !== "string" || sessionId === "") return null;
  if (typeof token !== "string" || token === "") return null;
  if (!isRecord(service)) return null;
  if (typeof service.name !== "string" || typeof service.version !== "string") return null;
  if (typeof service.protocol !== "string") return null;
  const read = readCapabilities(capabilities);
  if (read === null) return null;
  if (typeof openedAt !== "number" || !Number.isFinite(openedAt)) return null;
  if (typeof sessionsIssued !== "number" || !Number.isFinite(sessionsIssued)) return null;

  return {
    token,
    info: Object.freeze({
      sessionId,
      service: Object.freeze({ name: service.name, version: service.version, protocol: service.protocol }),
      capabilities: read,
      openedAt,
      sessionsIssued,
    }),
  };
}

/** Validates a health payload. */
export function readHealth(payload: unknown): LocalServiceHealth | null {
  if (!isRecord(payload)) return null;
  const { target, attached, capabilities, sessionId, attachedAt, sessionsIssued } = payload;
  if (!isRecord(target)) return null;
  if (!isBoolean(target.available) || !isBoolean(target.ready)) return null;
  if (typeof target.version !== "string" || target.version === "") return null;
  if (!isBoolean(attached)) return null;
  const read = readCapabilities(capabilities);
  if (read === null) return null;
  if (typeof sessionsIssued !== "number" || !Number.isFinite(sessionsIssued)) return null;

  return Object.freeze({
    target: Object.freeze({ available: target.available, ready: target.ready, version: target.version }),
    attached,
    capabilities: read,
    sessionId: typeof sessionId === "string" && sessionId !== "" ? sessionId : null,
    attachedAt: typeof attachedAt === "number" && Number.isFinite(attachedAt) ? attachedAt : null,
    sessionsIssued,
  });
}

/** Validates an attach payload. */
export function readAttached(payload: unknown): { attachedAt: number } | null {
  if (!isRecord(payload)) return null;
  if (payload.attached !== true) return null;
  const { attachedAt } = payload;
  if (typeof attachedAt !== "number" || !Number.isFinite(attachedAt)) return null;
  return { attachedAt };
}

/**
 * True when a refusal means the session this client holds is no longer the one
 * the service will act on. Every one of these is a lost session rather than a
 * failed operation, and they are the only refusals that end a session.
 */
export function isSessionLost(code: LocalServiceErrorCode): boolean {
  return code === "SESSION_STALE" || code === "AUTH_INVALID" || code === "AUTH_REQUIRED";
}

File: src-tauri/src/local_service/protocol.rs
//! The Nova local service wire protocol.
//!
//! This module is pure: it defines the request and response envelopes, the
//! error taxonomy and the limits, and it parses raw text into a validated
//! request. It holds no state, opens nothing and contacts nothing, so every
//! rule below is unit-tested without a Tauri application around it.
//!
//! The frontend's copy of these constants lives in
//! `src/features/backend/backends/localService/protocol.ts`. Nothing at runtime
//! could notice the two drifting apart — a mismatched version would simply
//! refuse every request — so `protocolParity.test.ts` reads this file and
//! compares them.

use serde::{Deserialize, Serialize};

/// Bumped only on an incompatible change; both sides must agree on it exactly.
pub const PROTOCOL_VERSION: &str = "NOVA_LOCAL_SERVICE_V1";

/// Largest message accepted in either direction, in bytes. An oversized message
/// is refused with a structured error rather than truncated or parsed.
pub const MAX_MESSAGE_BYTES: usize = 64 * 1024;

/// Largest request id accepted, in bytes.
///
/// The id is the one caller-supplied string the service echoes back verbatim,
/// so it is the one with a length of its own: a request that arrives with a
/// 60 KiB id is refused rather than answered with a 60 KiB response. Nova's own
/// ids are UUIDs, so nothing legitimate comes near this.
pub const MAX_REQUEST_ID_BYTES: usize = 128;

/// What the service calls itself where the UI names it.
pub const SERVICE_NAME: &str = "Nova Local Service";

/// The service build, reported in the handshake. Not the application version.
pub const SERVICE_VERSION: &str = "1";

/// The one capability this service implements, and so the only name a handshake
/// may require.
pub const CAPABILITY_TARGET: &str = "target";

/// Every operation the service answers. Anything else is refused by name.
pub const OPERATIONS: [&str; 5] = ["handshake", "health", "attach", "detach", "shutdown"];

/// Structured refusals. Every one of these is reachable, and each is produced by
/// exactly one rule; there is no catch-all code.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ErrorCode {
    /// The message is larger than [`MAX_MESSAGE_BYTES`].
    MessageTooLarge,
    /// The message is not JSON, or is not a request envelope.
    MessageMalformed,
    /// The message names a protocol version this service does not speak.
    ProtocolUnsupported,
    /// The operation needs a session and the message carries none.
    AuthRequired,
    /// The session named is the active one and the token does not match it.
    AuthInvalid,
    /// The session named is not the active one: it was superseded or closed.
    SessionStale,
    /// The handshake required a capability this service does not implement.
    CapabilityUnsupported,
    /// The operation is not one this service implements.
    OperationUnsupported,
    /// An attach arrived while the session was already attached.
    AlreadyAttached,
    /// The service could not do its part — today, only when the operating
    /// system refuses the randomness a session token is made of.
    ServiceUnavailable,
}

impl ErrorCode {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::MessageTooLarge => "MESSAGE_TOO_LARGE",
            Self::MessageMalformed => "MESSAGE_MALFORMED",
            Self::ProtocolUnsupported => "PROTOCOL_UNSUPPORTED",
            Self::AuthRequired => "AUTH_REQUIRED",
            Self::AuthInvalid => "AUTH_INVALID",
            Self::SessionStale => "SESSION_STALE",
            Self::CapabilityUnsupported => "CAPABILITY_UNSUPPORTED",
            Self::OperationUnsupported => "OPERATION_UNSUPPORTED",
            Self::AlreadyAttached => "ALREADY_ATTACHED",
            Self::ServiceUnavailable => "SERVICE_UNAVAILABLE",
        }
    }
}

/// A refusal, as it goes on the wire.
///
/// `details` carries technical context for a developer. It is built from what
/// the service knows about its own state and about the shape of the message —
/// never from a token, and never from a path.
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ProtocolError {
    pub code: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
}

impl ProtocolError {
    pub fn new(code: ErrorCode, message: &str) -> Self {
        Self {
            code: code.as_str().to_owned(),
            message: message.to_owned(),
            details: None,
        }
    }

    pub fn with_details(code: ErrorCode, message: &str, details: String) -> Self {
        Self {
            code: code.as_str().to_owned(),
            message: message.to_owned(),
            details: Some(details),
        }
    }
}

/// One request, as the frontend sends it.
///
/// `session_id` and `token` are absent for a handshake and required for
/// everything else. The service decides that, not serde, so a missing field is
/// refused with `AUTH_REQUIRED` rather than `MESSAGE_MALFORMED`.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Request {
    pub protocol: String,
    pub op: String,
    pub request_id: String,
    #[serde(default)]
    pub session_id: Option<String>,
    #[serde(default)]
    pub token: Option<String>,
    #[serde(default)]
    pub payload: serde_json::Value,
}

/// How a raw message was read.
#[derive(Debug)]
pub enum Decoded {
    Request(Box<Request>),
    Rejected {
        /// Recovered from the message when possible, so a refusal can still be
        /// correlated with the request that caused it; empty when it could not.
        request_id: String,
        error: ProtocolError,
    },
}

/// Pulls the request id out of a message that failed validation. Truncated,
/// because it is echoed back and nothing else validated it.
fn salvage_request_id(raw: &str) -> String {
    let Ok(value) = serde_json::from_str::<serde_json::Value>(raw) else {
        return String::new();
    };
    value
        .get("requestId")
        .and_then(serde_json::Value::as_str)
        .unwrap_or_default()
        .chars()
        .take(MAX_REQUEST_ID_BYTES)
        .collect()
}

/// Parses and validates a raw inbound message.
///
/// The order is load-bearing and mirrored on the frontend: size, then JSON,
/// then the envelope shape, then the protocol version. A message that is too
/// large is never parsed, and a version mismatch is reported distinctly from a
/// malformed one so the UI can explain which of the two happened.
pub fn decode(raw: &str) -> Decoded {
    if raw.len() > MAX_MESSAGE_BYTES {
        return Decoded::Rejected {
            request_id: String::new(),
            error: ProtocolError::with_details(
                ErrorCode::MessageTooLarge,
                "The message is larger than the local service accepts.",
                format!("{} bytes; the limit is {MAX_MESSAGE_BYTES}", raw.len()),
            ),
        };
    }

    let request: Request = match serde_json::from_str(raw) {
        Ok(request) => request,
        Err(error) => {
            return Decoded::Rejected {
                request_id: salvage_request_id(raw),
                error: ProtocolError::with_details(
                    ErrorCode::MessageMalformed,
                    "The message is not a local service request.",
                    format!("at line {}, column {}", error.line(), error.column()),
                ),
            };
        }
    };

    if request.request_id.is_empty() {
        return Decoded::Rejected {
            request_id: String::new(),
            error: ProtocolError::new(
                ErrorCode::MessageMalformed,
                "The message carries no request id.",
            ),
        };
    }

    if request.request_id.len() > MAX_REQUEST_ID_BYTES {
        return Decoded::Rejected {
            request_id: String::new(),
            error: ProtocolError::with_details(
                ErrorCode::MessageMalformed,
                "The message's request id is longer than the local service accepts.",
                format!(
                    "{} bytes; the limit is {MAX_REQUEST_ID_BYTES}",
                    request.request_id.len()
                ),
            ),
        };
    }

    if request.protocol != PROTOCOL_VERSION {
        return Decoded::Rejected {
            request_id: request.request_id.clone(),
            error: ProtocolError::with_details(
                ErrorCode::ProtocolUnsupported,
                "The message uses a protocol version this service does not speak.",
                format!(
                    "expected {PROTOCOL_VERSION}, received {}",
                    request.protocol.chars().take(64).collect::<String>()
                ),
            ),
        };
    }

    Decoded::Request(Box::new(request))
}

/// Serializes a successful answer. The payload is already the op's own shape.
pub fn ok_response(request_id: &str, payload: serde_json::Value) -> String {
    serde_json::json!({
        "protocol": PROTOCOL_VERSION,
        "requestId": request_id,
        "ok": true,
        "payload": payload,
    })
    .to_string()
}

/// Serializes a refusal.
pub fn error_response(request_id: &str, error: &ProtocolError) -> String {
    serde_json::json!({
        "protocol": PROTOCOL_VERSION,
        "requestId": request_id,
        "ok": false,
        "error": error,
    })
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn message(protocol: &str, op: &str) -> String {
        serde_json::json!({
            "protocol": protocol,
            "op": op,
            "requestId": "request-1",
        })
        .to_string()
    }

    fn rejection(raw: &str) -> ProtocolError {
        match decode(raw) {
            Decoded::Rejected { error, .. } => error,
            Decoded::Request(_) => panic!("expected {raw} to be refused"),
        }
    }

    #[test]
    fn a_well_formed_request_is_accepted() {
        let raw = message(PROTOCOL_VERSION, "health");
        match decode(&raw) {
            Decoded::Request(request) => {
                assert_eq!(request.op, "health");
                assert_eq!(request.request_id, "request-1");
                assert_eq!(request.session_id, None);
                assert_eq!(request.token, None);
            }
            Decoded::Rejected { error, .. } => panic!("refused a valid request: {error:?}"),
        }
    }

    #[test]
    fn an_oversized_message_is_refused_without_being_parsed() {
        let raw = "x".repeat(MAX_MESSAGE_BYTES + 1);
        assert_eq!(rejection(&raw).code, ErrorCode::MessageTooLarge.as_str());
    }

    #[test]
    fn a_message_within_the_limit_is_still_read() {
        let padding = MAX_MESSAGE_BYTES - 1024;
        let raw = serde_json::json!({
            "protocol": PROTOCOL_VERSION,
            "op": "health",
            "requestId": "request-1",
            "payload": { "pad": "y".repeat(padding) },
        })
        .to_string();
        assert!(raw.len() <= MAX_MESSAGE_BYTES, "{} bytes", raw.len());
        assert!(matches!(decode(&raw), Decoded::Request(_)));
    }

    #[test]
    fn malformed_messages_are_refused_by_shape() {
        for raw in [
            "",
            "{",
            "not json",
            "[]",
            "null",
            "{\"protocol\":1}",
            "{\"op\":\"health\",\"requestId\":\"r\"}",
            "{\"protocol\":\"NOVA_LOCAL_SERVICE_V1\",\"requestId\":\"r\"}",
        ] {
            assert_eq!(
                rejection(raw).code,
                ErrorCode::MessageMalformed.as_str(),
                "{raw}"
            );
        }
    }

    #[test]
    fn an_empty_request_id_is_malformed() {
        let raw = serde_json::json!({
            "protocol": PROTOCOL_VERSION,
            "op": "health",
            "requestId": "",
        })
        .to_string();
        assert_eq!(rejection(&raw).code, ErrorCode::MessageMalformed.as_str());
    }

    #[test]
    fn a_request_id_the_service_would_have_to_echo_back_is_bounded() {
        let raw = serde_json::json!({
            "protocol": PROTOCOL_VERSION,
            "op": "health",
            "requestId": "x".repeat(MAX_REQUEST_ID_BYTES + 1),
        })
        .to_string();
        assert!(
            raw.len() < MAX_MESSAGE_BYTES,
            "the message itself is not oversized"
        );

        match decode(&raw) {
            Decoded::Rejected { request_id, error } => {
                assert_eq!(error.code, ErrorCode::MessageMalformed.as_str());
                assert_eq!(request_id, "", "an id that long is not echoed back either");
            }
            Decoded::Request(_) => panic!("expected a refusal"),
        }

        // One byte shorter is ordinary.
        let allowed = serde_json::json!({
            "protocol": PROTOCOL_VERSION,
            "op": "health",
            "requestId": "x".repeat(MAX_REQUEST_ID_BYTES),
        })
        .to_string();
        assert!(matches!(decode(&allowed), Decoded::Request(_)));
    }

    #[test]
    fn another_protocol_version_is_refused_distinctly() {
        let error = rejection(&message("NOVA_LOCAL_SERVICE_V2", "health"));
        assert_eq!(error.code, ErrorCode::ProtocolUnsupported.as_str());
        assert!(
            error.details.unwrap_or_default().contains("V2"),
            "the refusal names the version it was given"
        );
    }

    #[test]
    fn a_refusal_still_names_the_request_it_answers() {
        let raw = "{\"protocol\":\"NOVA_LOCAL_SERVICE_V1\",\"requestId\":\"request-9\"}";
        match decode(raw) {
            Decoded::Rejected { request_id, .. } => assert_eq!(request_id, "request-9"),
            Decoded::Request(_) => panic!("expected a refusal"),
        }
    }

    #[test]
    fn responses_carry_the_protocol_and_the_request_id() {
        let ok = ok_response("request-1", serde_json::json!({ "attached": true }));
        assert!(ok.contains(PROTOCOL_VERSION));
        assert!(ok.contains("\"requestId\":\"request-1\""));
        assert!(ok.contains("\"ok\":true"));

        let refused = error_response(
            "request-2",
            &ProtocolError::new(ErrorCode::AuthRequired, "No session."),
        );
        assert!(refused.contains("\"ok\":false"));
        assert!(refused.contains("AUTH_REQUIRED"));
        assert!(!refused.contains("\"details\""), "no details, no field");
    }

    #[test]
    fn every_error_code_has_a_distinct_wire_name() {
        let codes = [
            ErrorCode::MessageTooLarge,
            ErrorCode::MessageMalformed,
            ErrorCode::ProtocolUnsupported,
            ErrorCode::AuthRequired,
            ErrorCode::AuthInvalid,
            ErrorCode::SessionStale,
            ErrorCode::CapabilityUnsupported,
            ErrorCode::OperationUnsupported,
            ErrorCode::AlreadyAttached,
            ErrorCode::ServiceUnavailable,
        ];
        let mut seen = std::collections::HashSet::new();
        for code in codes {
            assert!(
                seen.insert(code.as_str()),
                "{} is duplicated",
                code.as_str()
            );
            assert!(!code.as_str().is_empty());
        }
    }
}
